src/ext/random.c

```diff
@@ -5,24 +5,63 @@
 #include <ext/standard/php_rand.h>
 
 #include "configuration.h"
+#include "ddtrace.h"
 #include "third-party/mt19937-64.h"
 
-void dd_trace_seed_prng(TSRMLS_D) {
+ZEND_EXTERN_MODULE_GLOBALS(ddtrace);
+
+void ddtrace_seed_prng(TSRMLS_D) {
     if (get_dd_trace_debug_prng_seed() > 0) {
         init_genrand64((unsigned long long)get_dd_trace_debug_prng_seed());
     } else {
         init_genrand64((unsigned long long)GENERATE_SEED());
     }
 }
 
-static long long generate_id() {
+void ddtrace_init_span_id_stack(TSRMLS_D) {
+    DDTRACE_G(root_span_id) = 0;
+    DDTRACE_G(span_ids_top) = NULL;
+}
+
+void ddtrace_free_span_id_stack(TSRMLS_D) {
+    while (DDTRACE_G(span_ids_top) != NULL) {
+        ddtrace_span_ids_t *stack = DDTRACE_G(span_ids_top);
+        DDTRACE_G(span_ids_top) = stack->next;
+        efree(stack);
+    }
+}
+
+uint64_t ddtrace_push_span_id(TSRMLS_D) {
+    ddtrace_span_ids_t *stack = ecalloc(1, sizeof(ddtrace_span_ids_t));
     // We shift one bit to get 63-bit
-    return (long long)(genrand64_int64() >> 1);
+    stack->id = (uint64_t)(genrand64_int64() >> 1);
+    stack->next = DDTRACE_G(span_ids_top);
+    DDTRACE_G(span_ids_top) = stack;
+    // Assuming the first call this function is for the root span
+    if (DDTRACE_G(root_span_id) == 0) {
+        DDTRACE_G(root_span_id) = stack->id;
+    }
+    return stack->id;
 }
 
-#if PHP_VERSION_ID >= 70200
-// zend_strpprintf() wasn't exposed until PHP 7.2
-zend_string *dd_trace_generate_id() { return zend_strpprintf(0, "%llu", generate_id()); }
-#else
-void dd_trace_generate_id(char* buf) { php_sprintf(buf, "%llu", generate_id()); }
-#endif
+uint64_t ddtrace_pop_span_id(TSRMLS_D) {
+    if (DDTRACE_G(span_ids_top) == NULL) {
+        return 0;
```

Comment:
Yes, it's used to indicate a root span. Essentially if there's no active span and we get `0` here, we know that the span we're creating is a root span. This is helpful since we don't know if C or PHP has the root span. :)